include/seqan3/io/sequence_file/format_sam.hpp

```diff
@@ -0,0 +1,308 @@
+     * \{
+     */
+     //!\brief Default constructor is explicitly deleted, you need to give a stream or file name.
+    sequence_file_format_sam() = default;
+    //!\brief Copy construction is explicitly deleted, because you can't have multiple access to the same file.
+    sequence_file_format_sam(sequence_file_format_sam const &) = delete;
+    //!\brief Copy assignment is explicitly deleted, because you can't have multiple access to the same file.
+    sequence_file_format_sam & operator=(sequence_file_format_sam const &) = delete;
+    //!\brief Move construction is defaulted.
+    sequence_file_format_sam(sequence_file_format_sam &&) = default;
+    //!\brief Move assignment is defaulted.
+    sequence_file_format_sam & operator=(sequence_file_format_sam &&) = default;
+    //!\}
+
+    //!\brief The valid file extensions for this format; note that you can modify this value.
+    static inline std::vector<std::string> file_extensions
+    {
+        { "sam" },
+    };
+
+    //!\copydoc sequence_file_input_format_concept::read
+    template <typename stream_type,     // constraints checked by file
+              typename seq_legal_alph_type, bool seq_qual_combined,
+              typename seq_type,        // other constraints checked inside function
+              typename id_type,
+              typename qual_type>
+    void read(stream_type                     & stream,
+              sequence_file_input_options<seq_legal_alph_type, seq_qual_combined> const & options,
+              seq_type                        & sequence,
+              id_type                         & id,
+              qual_type                       & qualities)
+    {
+        auto stream_view = view::subrange<decltype(std::istreambuf_iterator<char>{stream}),
+                                          decltype(std::istreambuf_iterator<char>{})>
+                            {std::istreambuf_iterator<char>{stream},
+                             std::istreambuf_iterator<char>{}};
+
+        // cache the begin position so we write quals to the same position as seq in seq_qual case
+        size_t sequence_size_before, sequence_size_after = 0;
+        if constexpr (!detail::decays_to_ignore_v<seq_type>)
+            sequence_size_before = ranges::size(sequence);
+
+        auto constexpr is_tab = is_char<'\t'>;
+        auto constexpr is_cntrl= is_char<'\n'>;
+
+        while (is_char<'@'>(*ranges::begin(stream_view)))
+            detail::consume(stream_view | view::take_line);
+
+        /* ID */
+        std::string first; // Checking if id is given or set to *
+        ranges::copy(stream_view | view::take_until_or_throw(is_tab || is_blank), std::back_inserter(first));
+        if (first == "*")
+            throw parse_error{"The ID field may not be empty for SAM files."};
+
+        if constexpr (!detail::decays_to_ignore_v<id_type>)
+        {
+            ranges::copy(first | view::char_to<value_type_t<id_type>>, std::back_inserter(id)); //Add first to id
```

Comment:
Please do not use temporary strings. You can call begin() as often as you want without consuming the stream.
Replace the code by this
```c++
if(is_char<'*'>(*begin(stream_view)))
    throw parse_error{"The ID field may not be empty for SAM files."};

```